### PR TITLE
Fix some issues with very high drag

### LINKED ISSUE #4379: Extremely high drag causes unusual speed behaviour
With a high enough drag, thrusting opposite to current motion causes extremely high speeds to occur. 

This is reproducible (and easily viewable) with a bastion with 5000 drag otherwise stock.

0.9.9 steam version, win 10 1803 

Thanks to Xyma.FLIF on discord for spotting this.

### PR DESCRIPTION
**Pull request by warp-core**
**Bugfix:** Fixes #4379

## Fix Details
Effectively limits the drag value to not exceed the mass. This prevents the drag alone from completely negating the existing velocity, regardless of the applied acceleration.
This fixes the linked issue because it means that the very high drag will not completely negate the velocity and instead create an acceleration in the opposite direction that far exceeds the existing velocity.

Another change also prevents a separate bug where the velocity of a disabled ship with very high drag will become completely unbound.

## Testing Done
Fly around in a ship with very high drag and see that the velocity doesn't become unbound.
Also, disable a moving ship with very high drag and see that its velocity doesn't become unbound.

## Save File
This save file can be used to verify the bugfix. The bug will occur when using {{insert commit hash / version}}, and will not occur when using this branch's build.
[test test.txt](https://github.com/endless-sky/endless-sky/files/11435342/test.test.txt)

### CODE REVIEW

**Comment by TomGoodIdea** [comment]
Why is the limit applied in the Move function and not just in Drag() directly?

**Comment by warp-core** [comment]
> Why is the limit applied in the Move function and not just in Drag() directly?

I recall considering that but I don't remember why I didn't do it, so it probably wasn't important.

**Comment by TomGoodIdea** [comment]
Then I guess you should change this line too:
https://github.com/endless-sky/endless-sky/blob/755789be76e78fa17287befdae317bc4de3592e0/source/AI.cpp#L2330

**Comment by Ferociousfeind** [comment]
I feel like there's a better way to do this than to simply clamp drag...

**Comment by Ferociousfeind** [comment]
Here, I've crunched the numbers.

What the thrust/drag situation is emulating is the curve `f(t) = thrust / drag * (1 - (magic constant) ^ (t * drag))`. This is the "infinitely small physics step" version of the recursive function `f(n) = thrust + f(n-1) * (1 - drag)`, which is exactly what thrusting does- subtracts your current velocity * your drag, then adds your acceleration. These functions roughly approximate each other, but while the exponential function is well-behaved at all values for thrust and drag, the recursive function is badly behaved whenever drag is >= 1. See:
image
image
image

These are not very visible in-game because the drag value in the data files (which can get as high as, like, 24 on some freighters?) is divided by 100 (or 1000?) before being used (and so 24 becomes 0.24, or 0.024 drag), so the big issue is at 100 (or 1000) drag, and beyond there. But it, evidenced by this PR in the first place, DOES become an issue at some point.

I'd recommend, instead of capping drag arbitrarily, using the continuous form of the function for its well-behaved-ness. This isn't quite as hard as it seems, but it might be more performance-intensive than expected (though, drag doesn't usually change mid-flight, so the drag coefficient could be calculated on the fly. It's a magic constant raised to the power of the drag in the data files). 

This would be changing the line
```
Point dragAcceleration = acceleration - velocity * (Drag() / mass);
```
to
```
Point dragAcceleration = acceleration - velocity * (1. - pow(0.36787946, Drag() / mass));
```
Ultimately it breaks down to acceleration - velocity * some_constant_between_0_and_1, the important part being that "between 0 and 1" part being always true. As drag increases, this constant asymptotically, exponentially approaches 1 (i.e. very high drag means you are met with nearly 100% of your velocity as opposing acceleration). The 0.367 constant would need recalculating, depending on what amount of drag is 100% drag.

The constant should be such that, given thrust(a) and drag(d), the formula `y=(a/d)*(c^d)` and the formula `y=a` have a single intersection at (0, 0). This is equivalent to zero velocity allowing an instantaneous acceleration that is as if you have no drag:
image

If the constant is too large, then acceleration at 0 velocity will be too low:
image

If the constant is too small, then acceleration at 0 velocity will be too high:
image

And the important constraint, the `y=(a/d)*(c^d)` formula should have a horizontal asymptote at the ship's top speed for its thrust/drag. Quick testing seems as if the constant _should_ be 0.990049834, for where 100 drag is actually thrust == top speed, and for if 1000 drag makes thrust == top speed, the constant should be 0.99900049984 (But, of course, I'd want to inspect the code more closely to confirm if these are the correct values) (also at such sizes, I suspect floating point precision might cause some issues... perhaps manipulate the drag value and keep the constant closer to 0.5)

I don't know if I need to say more, tbh it's a little late, so I'll close this doctorate thesis by saying "please put it in da game, it behaves well"

**Comment by Ferociousfeind** [comment]
By the way

> also at such sizes, I suspect floating point precision might cause some issues... perhaps manipulate the drag value and keep the constant closer to 0.5

means "divide the data file drag amount by some number (685.145, via experimentation), and then raise 0.5 to that power"

**Comment by Ferociousfeind** [comment]
Also by the way, heat dissipation has the same problem and the same solution. It's why I remembered this  in the first place, same sort of question came up and I realized I could reuse my work here because drag = heat dissipation mechanically

**Comment by Ferociousfeind** [comment]
wtf who put conflicts on my branch

File: source/Ship.h [hunk 1]
```diff
@@ -346,8 +346,12 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
 	double MaximumHeat() const;
 	// Calculate the multiplier for cooling efficiency.
 	double CoolingEfficiency() const;
-	// Calculate the ship's drag after accounting for drag reduction.
+	// Calculate the drag on this ship. If the raw drag value is greater than the
+	// ship's mass, then the drag becomes the mass.
```
Comment (Amazinite) [inline source/Ship.h:350]:
```suggestion
	// Calculate the drag on this ship. The drag can be no greater than the mass.
```

File: source/Ship.cpp [hunk 2]
```diff
@@ -2661,10 +2661,24 @@ int Ship::Crew() const
 
 
 
-// Calculate drag, accounting for drag reduction.
+// Calculate the drag on this ship. If the raw drag value is greater than the
+// ship's mass, then the drag becomes the mass.
```
Comment (Amazinite) [inline source/Ship.cpp:2665]:
```suggestion
// Calculate the drag on this ship. The drag can be no greater than the mass.
```

**Pull request merged**